src/gluonts/dataset/loader.py

```diff
@@ -11,186 +11,270 @@
+                    self.base_iterator, self.shuffle_buffer_length
+                )
+            )
+
+        # If buffer still empty, means all elements used, return a signal of
+        # end of iterator
+        if not self.shuffle_buffer:
+            raise StopIteration
+
+        # Choose an element at a random index and yield it and fill it with
+        # the next element in the sequential generator
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        next_sample = self.shuffle_buffer[idx]
+
+        # Replace the index with the next element in the iterator if the
+        # iterator has not finished. Delete the index otherwise.
+        try:
+            self.shuffle_buffer[idx] = next(self.base_iterator)
+        except StopIteration:
+            del self.shuffle_buffer[idx]
+
+        return next_sample
+
+
+class MultiProcessBatcher(Iterator):
     def __init__(
         self,
-        dataset: Dataset,
-        *,
-        transform: Transformation,
-        cyclic: bool,
-        is_train: bool,
+        base_iterable: Iterable,
         batch_size: int,
-        ctx: mx.Context,
-        dtype: DType = np.float32,
-        num_workers: Optional[int] = None,
-        num_prefetch: Optional[int] = None,
-        shuffle_buffer_length: Optional[int] = None,
-        **kwargs,
-    ) -> None:
+        batchify_fn: Callable,
+        num_workers: int,
+        max_queue_size: Optional[int] = None,
+        todevice_fn: Callable = lambda x: x,
+    ):
+        assert num_workers >= 1
+        assert max_queue_size is None or max_queue_size >= num_workers
+
+        self.base_iterable = base_iterable
         self.batch_size = batch_size
-        self.ctx = ctx
-        self.dtype = dtype
-        self.is_train = is_train
-        self.transform = transform
-        self.cyclic = cyclic
-        self.logger = logging.getLogger(__name__)
+        self.batchify_fn = batchify_fn
+        self.todevice_fn = todevice_fn
         self.num_workers = num_workers
-        self.num_prefetch = num_prefetch
-        self.shuffle_buffer_length = shuffle_buffer_length
-
-        self.parallel_data_loader = ParallelDataLoader(
-            dataset=dataset,
-            transformation=self.transform,
-            cyclic=self.cyclic,
-            is_train=self.is_train,
-            batch_size=self.batch_size,
-            ctx=self.ctx,
-            dtype=self.dtype,
-            num_workers=self.num_workers,
-            num_prefetch=self.num_prefetch,
-            shuffle_buffer_length=self.shuffle_buffer_length,
-            **kwargs,
+        self.max_queue_size = (
+            max_queue_size if max_queue_size is not None else 5 * num_workers
         )
 
-    def __iter__(self) -> Iterator[DataBatch]:
-        # Will take all batches, so that all data is sampled exactly once
-        yield from self.parallel_data_loader
+        self.manager = Manager()
+        self.data_queue = self.manager.Queue(maxsize=self.max_queue_size)
```

Comment:
`data_queue` is not really a helpful name. Is this the input or the output queue?

Reply:
depends who you ask: is the output queue for the workers, it's the input queue for the main process

Reply:
Not sure I can follow that logic. Can we name it `result_queue` or something similar?